assets/src/components/detours/detourMap.tsx

```diff
@@ -0,0 +1,203 @@
+  const [endPoint, setEndPoint] = useState<LatLngExpression | null>(null)
+  const [detourPositions, setDetourPositions] = useState<LatLngExpression[]>([])
+
+  const onAddDetourPosition = (p: LatLngExpression) =>
+    setDetourPositions((positions) => [...positions, p])
+
+  return (
+    <Map vehicles={[]}>
+      <CustomControl position="topleft" className="leaflet-bar">
+        <Button
+          variant="primary"
+          disabled={
+            startPoint === null ||
+            endPoint !== null ||
+            detourPositions.length === 1
+          }
+          onClick={() =>
+            setDetourPositions((positions) =>
+              positions.slice(0, positions.length - 1)
+            )
+          }
+        >
+          Clear Last Waypoint
+        </Button>
+      </CustomControl>
+      <RouteShapeWithDetour
+        originalShape={shape}
+        startPoint={startPoint}
+        onSetStartPoint={setStartPoint}
+        endPoint={endPoint}
+        onSetEndPoint={setEndPoint}
+        detourPositions={detourPositions}
+        onAddDetourPosition={onAddDetourPosition}
+      />
+    </Map>
+  )
+}
+
+const RouteShapeWithDetour = ({
+  originalShape,
+  startPoint,
+  onSetStartPoint,
+  endPoint,
+  onSetEndPoint,
+  detourPositions,
+  onAddDetourPosition,
+}: {
+  originalShape: Shape
+  startPoint: LatLngExpression | null
+  onSetStartPoint: (p: LatLngExpression | null) => void
+  endPoint: LatLngExpression | null
+  onSetEndPoint: (p: LatLngExpression | null) => void
+  detourPositions: LatLngExpression[]
+  onAddDetourPosition: (p: LatLngExpression) => void
+}) => {
+  const routeShapePositions: LatLngExpression[] = originalShape.points.map(
+    (point) => [point.lat, point.lon]
+  )
+
+  const map = useMap()
+
+  useMapEvent("click", (e) => {
+    if (startPoint !== null && endPoint === null) {
+      onAddDetourPosition(e.latlng)
+    }
+  })
+
+  // points on the detour not already represented by the start and end
+  const uniqueDetourPositions =
+    detourPositions.length === 0
+      ? []
+      : endPoint === null
+      ? detourPositions.slice(1)
+      : detourPositions.slice(1, -2)
+
+  return (
+    <>
+      <Polyline
+        positions={routeShapePositions}
+        className="c-detour_map--original-route-shape"
+        eventHandlers={{
+          click: (e) => {
+            if (startPoint === null) {
+              const position = closestPosition(
+                routeShapePositions,
+                e.latlng,
+                map
+              )
+              onSetStartPoint(position)
+              position && onAddDetourPosition(position)
+            } else if (endPoint === null) {
+              const position = closestPosition(
+                routeShapePositions,
+                e.latlng,
+                map
+              )
+              onSetEndPoint(position)
+              position && onAddDetourPosition(position)
```

Comment:
This causes `onAddDetourPosition` to get called twice for the `endPoint`, since the `useMapEvent` handler also gets called for the same point. 
```suggestion
```

Reply:
It's not quite the same - in the `useMapEvent` handler, it directly adds the corresponding position the user clicked on the map. In this code, we specifically take the closest point on the shape first. So actually you're right that there's a problem here with it firing twice, but it's actually the other instance of it firing that we want to remove. I don't immediately see an obvious answer and just doing `e.originalEvent.stopPropagation()` didn't seem to do the trick - I'll have to keep digging.

Reply:
It might also not be a huge deal, except that it'll make some silly turn-by-turn directions.

Reply:
7bf105e2 (thanks @firestack!)